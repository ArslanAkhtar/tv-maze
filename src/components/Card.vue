<script setup lang="ts">
import { router } from '@/router'
import type { Show } from '@/helpers/types'
const props = defineProps<{
  item?: Show
}>()

const year = props.item?.premiered?.split('-')[0] || ''
const openDetails = (id: number | undefined) => {
  router.push({
    name: 'show',
    params: { id }
  })
}
</script>

<template>
  <v-card class="mx-auto rounded-lg movie-card">
    <v-img
      @click="() => openDetails(item?.id)"
      class="align-end text-white"
      :src="item?.image?.medium"
      cover
    />
    <div class="rating" v-if="item?.rating.average">
      <v-icon size="small" icon="mdi-star"></v-icon> {{ item?.rating.average }}
    </div>
    <v-card-title class="text-sm card-title"
      >{{ item?.name }} {{ year ? '( ' + year + ' )' : '' }}</v-card-title
    >
    <div class="genre mx-3 mb-3">
      <v-chip
        v-for="genre in item?.genres"
        :key="genre"
        class="ma-1"
        text-color="white"
        size="small"
      >
        {{ genre }}
      </v-chip>
    </div>
  </v-card>
</template>

<style scoped>
.movie-card {
  position: relative;
  transition: 0.1s all ease-in-out;
  height: 500px;
}
@media (max-width: 1280px) {
  .movie-card {
    height: 400px;
  }
}
@media (max-width: 960px) {
  .movie-card {
    height: unset !important;
  }
}
/* .movie-card :hover {
  transform: scale(0.5);
} */
.movie-card .rating {
  position: absolute;
  border-radius: 8px;
  top: 8px;
  right: 8px;
  background: rgba(0, 0, 0, 0.7);
  padding: 6px 12px;
  color: goldenrod;
  display: flex;
  align-items: center;
  gap: 5px;
}
.movie-card .v-img {
  cursor: pointer;
}
.movie-card .v-card-title {
  font-size: 18px !important;
}
</style>
