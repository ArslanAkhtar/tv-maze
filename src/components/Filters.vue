<template>
	<v-row justify="center" align="center" class="mt-3">
		<v-col cols="12" sm="6" md="4" xl="2">
			<v-text-field
				outlined
				color="yellow"
				density="compact"
				@update:modelValue="
					(e) => {
						$emit('search-change', e);
					}
				"
				@clearable="false"
				variant="solo"
				label="Search Shows"
				append-inner-icon="mdi-magnify"
			></v-text-field>
		</v-col>
		<v-col cols="12" sm="6" md="4" xl="2">
			<v-select
				density="compact"
				color="yellow"
				:items="genreFilters"
				label="Genres"
				clearable
				clear-icon="mdi-close-circle"
				multiple
				variant="solo"
				@update:modelValue="
					(genres) => {
						$emit('genre-change', genres);
					}
				"
			></v-select>
		</v-col>
		<v-col cols="12" sm="6" md="4" xl="2">
			<v-select
				density="compact"
				color="yellow"
				clearable
				clear-icon="mdi-close-circle"
				:items="rating"
				label="Rating"
				variant="solo"
				@update:modelValue="
					(rating) => {
						$emit('rating-change', rating);
					}
				"
			></v-select>
		</v-col>
	</v-row>
</template>
<script lang="ts" setup>
	defineProps(["genreFilters"]);
	defineEmits(["search-change", "genre-change", "rating-change"]);
	const rating = Array(9)
		.fill(1)
		.map((item: number, index: number) => `${index + 1} +`);
</script>
