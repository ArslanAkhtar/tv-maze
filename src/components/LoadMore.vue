<template>
	<div class="loadMore">
		<v-progress-circular
			indeterminate
			model-value="20"
			:size="51"
			:width="7"
		></v-progress-circular>
	</div>
</template>
