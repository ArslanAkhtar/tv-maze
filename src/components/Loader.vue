<script setup lang="ts">
// import { FulfillingBouncingCircleSpinner } from 'epic-spinners'
</script>
<template>
  <div class="loading">
    <v-progress-circular indeterminate model-value="20" :size="51" :width="7"></v-progress-circular>
  </div>
</template>
