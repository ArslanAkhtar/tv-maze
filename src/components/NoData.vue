<script setup lang="ts">
	defineProps({
		text: {
			type: String,
			required: true,
		},
	});
</script>

<template>
	<div class="no-data-found">
		<img alt="Vue logo" class="logo" src="@/assets/no-data.svg" width="320" />
		<div class="text-h4 font-weight-bold">{{ text }}</div>
	</div>
</template>
