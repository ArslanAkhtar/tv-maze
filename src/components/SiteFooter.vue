<template>
  <v-footer class="text-center d-flex flex-column">
    <div>
      <v-btn
        v-for="icon in icons"
        :key="icon.name"
        class="mx-4"
        :icon="icon.name"
        @click="openExternalLink(icon.link)"
        variant="text"
      ></v-btn>
    </div>

    <div class="pt-0">
      Phasellus feugiat arcu sapien, et iaculis ipsum elementum sit amet. Mauris cursus commodo
      interdum. Praesent ut risus eget metus luctus accumsan id ultrices nunc. Sed at orci sed massa
      consectetur dignissim a sit amet dui. Duis commodo vitae velit et faucibus. Morbi vehicula
      lacinia malesuada. Nulla placerat augue vel ipsum ultrices, cursus iaculis dui sollicitudin.
      Vestibulum eu ipsum vel diam elementum tempor vel ut orci. Orci varius natoque penatibus et
      magnis dis parturient montes, nascetur ridiculus mus.
    </div>

    <v-divider></v-divider>

    <div class="copyright">{{ new Date().getFullYear() }} â€” <strong>Muhammad Arslan</strong></div>
  </v-footer>
</template>

<script setup lang="ts">
const icons = [
  {
    name: 'mdi-web',
    link: 'https://www.muhammadarslan.nl/#home'
  },
  {
    name: 'mdi-linkedin',
    link: 'https://www.linkedin.com/in/arslan-akhtar'
  },
  {
    name: 'mdi-github',
    link: 'https://github.com/ArslanAkhtar'
  }
]

const openExternalLink = (url: string) => {
  window.open(url, '_blank')
}
</script>
