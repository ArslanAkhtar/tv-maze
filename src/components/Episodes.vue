<script setup lang="ts">
	import useFetchShows from "@/composables/FetchData";
	import Loader from "@/components/Loader.vue";
	import { imgURL, openExternalURL } from "@/helpers/utility";
	import Error from "@/components/Error.vue";

	const props = defineProps({
		id: {
			type: Number,
			required: true,
		},
	});

	const { loadShowEpisodes } = useFetchShows();
	const {
		data: episodes,
		isFetching,
		error: ErrorText,
	} = loadShowEpisodes(props.id);
</script>
<template>
	<v-list :height="300">
		<v-list-item
			v-if="episodes"
			v-for="(episode, i) in episodes"
			:key="i"
			@click="() => openExternalURL(episode.url)"
			:prepend-avatar="imgURL(episode.image?.medium)"
			class="pa-5"
			elevation="1"
		>
			<template v-if="episode.name" v-slot:title>
				{{ episode.name }}
			</template>
			<template v-if="episode.number" v-slot:subtitle>
				Episode {{ episode.number }}
			</template>
			<template v-slot:append>
				<v-icon>mdi-play</v-icon>
			</template>
		</v-list-item>
		<Loader v-else-if="isFetching" />
		<div class="d-flex justify-center align-center" style="height: 100%" v-else>
			<Error :text="ErrorText" />
		</div>
	</v-list>
</template>
