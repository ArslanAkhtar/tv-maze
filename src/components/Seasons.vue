<script setup lang="ts">
	import Episodes from "@/components/Episodes.vue";
	import type { Season } from "@/helpers/types";
	import type { PropType } from "vue";

	defineProps({
		seasons: {
			type: Array as PropType<Season[]>,
			required: true,
		},
	});
</script>

<template>
	<div class="seasons" v-if="seasons">
		<div class="text-h5 mb-5 font-weight-bold">Seasons:</div>
		<v-expansion-panels>
			<v-expansion-panel
				v-for="season in seasons"
				:key="season.id"
				:title="'Season ' + season.number"
				expand-icon="mdi-chevron-down"
				collapse-icon="mdi-chevron-up"
			>
				<v-expansion-panel-text>
					<Episodes :id="season.id" />
				</v-expansion-panel-text>
			</v-expansion-panel>
		</v-expansion-panels>
	</div>
</template>
