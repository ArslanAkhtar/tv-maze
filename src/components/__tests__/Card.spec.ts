import { describe, it, expect } from "vitest";
import { mount, shallowMount } from "@vue/test-utils";
import Card from "../Card.vue";
import { vuetify } from "@/main";
import * as components from "vuetify/components";
import mockShow from "./mockData/mockShow.json";
import { router } from "@/router";

describe("Card Component", () => {
	it("renders props", () => {
		const wrapper = mount(Card, { props: { item: mockShow } });
		expect(wrapper.props().item).toMatchObject(mockShow);
	});

	it("renders show name", () => {
		const wrapper = mount(Card, {
			global: {
				plugins: [vuetify],
			},
			props: { item: mockShow },
		});
		expect(wrapper.html().includes(mockShow.name)).toBe(true);
	});

	it("renders show genre", () => {
		const wrapper = mount(Card, {
			global: {
				plugins: [vuetify],
			},
			props: { item: mockShow },
			components: components,
		});
		expect(wrapper.html().includes(mockShow.genres[0])).toBe(true);
	});

	test("renders correctly", () => {
		const wrapper = shallowMount(Card);
		expect(wrapper.element).toMatchSnapshot();
	});

	test("openDetails navigates to the correct route", async () => {
		const wrapper = mount(Card, {
			global: {
				plugins: [router],
			},
		});
		const id = 1;
		await wrapper.vm.openDetails(id);
		expect(router.currentRoute.value.name).toBe("show");
		expect(router.currentRoute.value.params).toEqual({ id: "1" });
	});
	test("openDetails return if id is undefined", async () => {
		const wrapper = mount(Card, {
			global: {
				plugins: [router],
			},
		});
		const id = undefined;
		const res = await wrapper.vm.openDetails(id);
		expect(res).toEqual(null);
	});
});
