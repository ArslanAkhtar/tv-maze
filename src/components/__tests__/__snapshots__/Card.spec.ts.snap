// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card Component > renders correctly 1`] = `
<v-card
  class="mx-auto rounded-lg movie-card"
  data-v-c6c3362a=""
>
  <v-img
    class="align-end text-white"
    cover=""
    data-v-c6c3362a=""
  />
  <!--v-if-->
  <v-card-title
    class="text-sm card-title"
    data-v-c6c3362a=""
  >
     
  </v-card-title>
  <div
    class="genre mx-3 mb-3"
    data-v-c6c3362a=""
  >
    
    
  </div>
</v-card>
`;
